### PR TITLE
fetchgit: give output a nicer name

### PR DESCRIPTION
**Pull request by shlevy**

### CODE REVIEW

**Comment by vcunat** [comment]
- I tested this by building proot, but it doesn't work (and it builds on master).
- Dependency on nix-1.8 seems fine for master now, although we might prefer to keep that requirement only for fetchgit ATM (not whole nixpkgs).
- It might be nice to extend the name by a dash and the first few characters from ``rev``.

**Comment by shlevy** [comment]
Build of `proot.src` worked fine. Testing `proot` now...

**Comment by shlevy** [comment]
Re: nix requirement, we have a policy of allowing a dependency on the nix version used in the most recent stable branch I thought. But we can limit it to fetchgit if you want.

**Comment by shlevy** [comment]
`proot` build failure is due to hard-coding `git-export`, not a problem with this patch. I'll fix it and update the PR.

**Comment by shlevy** [comment]
@vcunat concerns addressed.

**Comment by shlevy** [comment]
rebased

**Comment by vcunat** [comment]
Ah, I see. I should've checked. More packages might have the same problem, but that should be easy to fix once Hydra uncovers the problems.

**Comment by iElectric** [comment]
```
$ git grep git-export
pkgs/applications/networking/instant-messengers/freetalk/default.nix:    name = "git-export-${s.name}";
pkgs/build-support/builder-defs/builder-defs.nix:                else if hasSuffixHack "-git-export" s then "git-dir"
pkgs/build-support/fetchgit/default.nix:, name ? "git-export"
pkgs/build-support/fetchgit/nix-prefetch-git:        finalPath=$(nix-store --print-fixed-path --recursive "$hashType" "$expHash" git-export)
pkgs/build-support/fetchgit/nix-prefetch-git:        tmpFile="$tmpPath/git-export"
pkgs/games/dwarf-fortress/default.nix:  sourceRoot = "git-export";
pkgs/games/dwarf-fortress/default.nix:    cp -f ./git-export/build/libgraphics.so $out/share/df_linux/libs/libgraphics.so
pkgs/tools/misc/ipxe/default.nix:  sourceRoot = "git-export/src";
pkgs/tools/misc/ldapvi/default.nix:    sourceRoot=git-export/ldapvi
pkgs/tools/misc/mdbtools/git.nix:    name = "mdbtools-git-export";
pkgs/tools/misc/pipelight/default.nix:  chmod u+w $wineDir/../git-export/debian/tools/
pkgs/tools/misc/pipelight/default.nix:        patchShebangs $wineDir/../git-export/debian/tools/
pkgs/tools/misc/pipelight/default.nix:        chmod -R +rwx ../git-export/
pkgs/tools/misc/pipelight/default.nix:        make -C ../git-export/patches DESTDIR=$wineDir install
pkgs/tools/system/proot/default.nix:  sourceRoot = "git-export/src";
```

File: pkgs/build-support/fetchgit/default.nix [hunk 1]
```diff
@@ -1,6 +1,20 @@
-{stdenv, git, cacert}:
-{url, rev ? "HEAD", md5 ? "", sha256 ? "", leaveDotGit ? false, fetchSubmodules ? true
-, name ? "git-export"
+{stdenv, git, cacert}: let
+  urlToName = url: rev: let
+    base = baseNameOf url;
+
+    matched = (builtins.match or (x: y:
```
Comment (edolstra) [inline pkgs/build-support/fetchgit/default.nix]: `match` is an undocumented function, please don't use it in Nixpkgs.

**Comment by edolstra** [comment]
Why is it important to have a "nice" name of fetchgit output? It's typically only used as input to other derivations, so not something that end users see.

**Comment by 7c6f434c** [comment]
>Why is it important to have a "nice" name of fetchgit output? It's typically only used as input to other derivations, so not something that end users see.

I'd say that it makes debugging less annoying when you look at an issue
created by interaction between updates of two or more upstream repos.

I have no opinion on whether this use case is worth the change.

**Comment by bjornfor** [comment]
I feel like an end user when I run nix-build or nixos-rebuild and see the list of store paths that will be created. Seing multiple "git-export" store paths gives little value, so +1 for changing to something descriptive, like fetcurl does.

**Comment by edolstra** [comment]
Yeah, that's a good point.

File: pkgs/build-support/fetchgit/default.nix (hunk 1 above)
Comment (shlevy) [inline pkgs/build-support/fetchgit/default.nix]: Why can't we just document it? It's pretty useful and it's not unsafe...

**Comment by vcunat** [comment]
Fixed packages referring to "git-export", tested them, and pushed.

TODO: it might be suitable to adapt nix-prefetch-git.

**Comment by vcunat** [comment]
It seems this caused some regressions, e.g. http://hydra.nixos.org/build/18716902. I'll probably look into them this evening or tomorrow.

**Comment by vcunat** [comment]
It's confirmed on mailing-list that this broke the nix-prefetch-git workflow.

**Comment by bjornfor** [comment]
It fails on typical fetchgitPrivate use:

```
error: while evaluating the attribute 'src' of the derivation "mypackage" at ...
while evaluating the derivation attribute 'name' at .../build-support/fetchgit/default.nix:46:10:
while evaluating 'urlToName' at .../build-support/fetchgit/default.nix:2:20, called from .../build-support/fetchgit/default.nix:17:10:
invalid file name 'git@server:repo.git'
```

**Comment by bjornfor** [comment]
Ah, it's the baseNameOf primop that doesn't like paths without a "/" in them:

```
nix-repl> baseNameOf "foo"
error: invalid file name ‘foo’
```

**Comment by volth** [comment]
It would be nice to handle ```fetchFromGitHub``` (```fetchFromBitbucket```, ...) as well

[They default ```name = "source"```](https://github.com/NixOS/nixpkgs/blob/05fedec/pkgs/top-level/all-packages.nix#L235):

**Comment by orivej** [comment]
@volth See https://github.com/NixOS/nixpkgs/commit/c3255fe8ec326d2c8fe9462d49ed83aa64d3e68f and https://github.com/NixOS/nixpkgs/pull/31004#issuecomment-340796748

**Pull request merged**